nengo_loihi/neurons.py

```diff
@@ -197,3 +269,354 @@ def nengo_build_nif_rate(model, nif_rate, neurons):
+            """Create a NeuronBuilder for the provided ops.
+
+            If all neurons share the same noise model, this will return a
+            subclass of NeuronBuilder. Otherwise, it will return a
+            NeuronBuilder instance to manage all noise models.
+            """
+            models = [getattr(op.neurons, 'nengo_dl_noise', None)
+                      for op in ops]
+            model_type = type(models[0]) if len(models) > 0 else None
+            equal_types = all(type(m) is model_type for m in models)
+
+            if equal_types and model_type in cls.builders:
+                return cls.builders[model_type](ops, signals, config, models)
+            else:
+                return NoiseBuilder(ops, signals, config, models)
+
+        def generate(self, period, tau_rc=None):
+            """Generate TensorFlow code to implement these noise models.
+
+            Parameters
+            ----------
+            period : tf.Tensor
+                The inter-spike periods of the neurons to add noise to.
+            tau_rc : tf.Tensor
+                The membrane time constant of the neurons (used by some noise
+                models).
+            """
+            raise NotImplementedError(
+                "Multiple noise models not supported for the same neuron type")
+
+    class NoNoiseBuilder(NoiseBuilder):
+        """nengo_dl builder for if there is no noise model."""
+
+        def generate(self, period, tau_rc=None):
+            return tf.reciprocal(period)
+
+    class RCNoiseBuilder(NoiseBuilder):
+        """Base class for noise models that use the neuron tau_rc."""
+
+        def __init__(self, ops, signals, config, models):
+            super(RCNoiseBuilder, self).__init__(
+                ops, signals, config, models)
+
+            tau_s = np.concatenate([
+                model.tau_s * one
+                for model, one in zip(self.models, self.np_ones)])
+            self.tau_s = signals.constant(tau_s, dtype=self.dtype)
+
+        @classmethod
+        def tensorflow(cls, period, tau_s, tau_rc):
+            """Generate TensorFlow code for this model type given parameters.
+
+            Parameters
+            ----------
+            period : tf.Tensor
+                The inter-spike periods of the neurons to add noise to.
+            tau_s : tf.Tensor
+                The time constant of the Lowpass synaptic filter.
+            tau_rc : tf.Tensor
+                The membrane time constant of the neurons (used by some noise
+                models).
+            """
+            raise NotImplementedError("Subclass must implement")
+
+        def generate(self, period, tau_rc=None):
+            assert tau_rc is not None
+            return self.tensorflow(period, self.tau_s, tau_rc)
+
+    class LowpassRCNoiseBuilder(RCNoiseBuilder):
+        """nengo_dl builder for the LowpassRCNoise model."""
+
+        @classmethod
+        def tensorflow(cls, period, tau_s, tau_rc):
```

Comment:
There was a docstring under the `RCNoiseBuilder` class, which fa90c22 gets rid of. In fact, it looks like this function no longer exists at all.

The reason that I had it was that originally, I was hoping to have a noise class that could do many different noise models together, and this kind of general function might be useful. But that's not happening anymore, so I think I agree that it makes sense to get rid of it.

Reply:
The docstring there was just "Generate TensorFlow code for this model type given parameters.", I was thinking more specifically about understanding the noise models here (e.g., what distribution is this function implementing, and why that distribution?).  For example, right now I can see all the math that is being implemented, but I can't see the reasoning underlying that math.  Basically a (very brief) introduction to why and how you developed these noise models, so that another developer coming along in the future could conceivably understand the "why" of the code without having to talk to you.  Maybe include a reference to the relevant section of your thesis?

Reply:
So should that docstring go under the noise model, or the noise model builder? Or maybe a bit of both? I definitely think that the noise model should have a bit more description and thesis reference. If I do that, do we still want something for the noise model builders?